### PR TITLE
Add slash in Debit card expiration

### PR DESCRIPTION
**Pull request by mananjadhav**
### Details

- Adding slash "/" in DebitCard expiry

### Fixed Issues

$ https://github.com/Expensify/App/issues/6891

### Tests
1. Test auto slash for 2 digits
2. Test backspace for 3 digits, 2 digits, digits ending with slash
3. Test with paste text
4. Tested with text already containing a slash
5. Tested for 5 digit and 7 digit text

- [x] Verify that no errors appear in the JS console
#### Contributor (PR Author) Checklist
- [x] I verified the PR has a small number of commits behind `main`
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I clearly indicated the environment tests should be run in (Staging vs Production)
- [x] I wrote testing steps that cover success & fail scenarios (if applicable)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests & verify they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors related to changes in this PR
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I added comments when the code was not self explanatory
    - [x] I put all copy / text shown in the product in all `src/languages/*` files (if applicable)
    - [x] I followed proper naming convention for platform-specific files (if applicable)
    - [x] I followed style guidelines (in [`Styling.md`](https://github.com/Expensify/App/blob/main/STYLING.md)) for all style edits I made
    - [x] I followed the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs))
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I corroborated the UI performance was not affected (the performance is the same than `main` branch)
- [x] If I created a new component I verified that a similar component doesn't exist in the codebase

#### PR Reviewer Checklist
- [ ] I verified the PR has a small number of commits behind `main`
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the testing environment is mentioned in the test steps
- [ ] I verified testing steps cover success & fail scenarios (if applicable)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors related to changes in this PR
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified comments were added when the code was not self explanatory
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files (if applicable)
    - [ ] I verified proper naming convention for platform-specific files was followed (if applicable)
    - [ ] I verified [style guidelines](https://github.com/Expensify/App/blob/main/STYLING.md) were followed
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components are not impacted by changes in this PR (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified the UI performance was not affected (the performance is the same than `main` branch)
- [ ] If a new component is created I verified that a similar component doesn't exist in the codebase

### QA Steps
1. Go to the Settings by clicking Profile Image on LHN
2. Select "Payments" from the Sidebar
3. Select Add Payment and choose "Debit Card"
4. Start typing the expiry date, it should automatically add slash(/) when you type 2 digits.
5. If you press backspace, then the slash should be removed automatically when you delete the third digit.

- [x] Verify that no errors appear in the JS console

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

web-slash-in-expiry-date

#### Mobile Web

mweb-slash-in-expiry-date

#### Desktop

desktop-slash-in-expiry-date

#### iOS

https://user-images.githubusercontent.com/3069065/158234529-f62814f1-63d0-48da-b6bb-6dec6ebd74db.mov

#### Android

https://user-images.githubusercontent.com/3069065/158236389-8e285a44-ca6a-4af5-a1b5-66c561370aa6.mov

### CODE REVIEW

**Comment by mananjadhav** [comment]
@parasharrajat @Julesssss  Did we check #7318 is working fine for Android and iOS? For me it doesn't text on prop change for native devices.

**Comment by parasharrajat** [comment]
> For me it doesn't text on prop change for native devices.

I didn't get it.

**Comment by parasharrajat** [comment]
What exactly are you facing?

**Comment by mananjadhav** [comment]
> What exactly are you facing?

The state value is updated in the form, but the `TextInput` doesn't show the updated value. Basically, when I do `Backspace`, it should delete the slash + digit and update the value but it doesn't reflect in `TextInput`. This happens only in native devices, works fine on Web, MWeb and Desktop.

**Comment by mananjadhav** [comment]
@parasharrajat  ^

**Comment by parasharrajat** [comment]
I will pull your branch see what is happening.

**Comment by parasharrajat** [comment]
I tested and it seems `this.state.expirationDate` is always empty. Could you please test it carefully?

**Comment by parasharrajat** [comment]
I tested on web.

**Comment by mananjadhav** [comment]
> I tested on web.

🤔  Here's the video of Web and I've also logged the `state.expirationDate` in the render. Can you take a pull again? Maybe I didn't push all changes?

https://user-images.githubusercontent.com/3069065/157687781-742748ab-e457-4648-b6d3-147021bfe3d4.mp4

**Comment by mananjadhav** [comment]
@parasharrajat I've just merged main to my branch. I've added some commented logs for your review when I was testing it. 

Also attaching the video for `iOS`. It has logs when it is just about to update, then in the render with updated state and in BaseTextInput, componentDidUpdate. If you see `this.props.value` has the latest updated value but `setNativeProps` doesn't update the value.

https://user-images.githubusercontent.com/3069065/157714818-e289c839-45a7-493d-b2e8-adca764c85ce.mp4

**Comment by parasharrajat** [comment]
Thanks for the Vid. I will test now.

**Comment by parasharrajat** [comment]
Ok, I see. It is due to that the TextInput internally processes the backspace and clears the last char. You tried to prevent it via `preventDefault()` I don't think it works on native.  

So this is the correct flow
1. keypress on native is fired before on changeText . 
2. When it does you manipulate the value and send an update to TextInput for new value on the same Tick. 
2. Now due to that ComponenDidUpdate trigger a native pop setting via `setnativeProps` (remember this does not care of React lifecycles). 
3. Now onChangetext is fired (TextInput internally handles the backspace and set the value which should be there when one char is removed). E.g. value 22/2 You press backspace value should be 22/ . 


I tried to use `this.allowExpirationDateChange` flag which works on Web due to preventDefault(). I suggest to move the whole logic to OnChangeText handler and then manage removal and addition.

**Comment by mananjadhav** [comment]
The `preventDefault` handling I added just yesterday. Earlier I was relying on the backspace handling and just taking the substring as `expirationDate.length-1`. So that the character removes the slash and character as well.

**Comment by parasharrajat** [comment]
Let me know your conclusions and how can I help.

**Comment by parasharrajat** [comment]
Good to hear it.

File: src/pages/settings/Payments/AddDebitCardPage.js [hunk 1]
```diff
@@ -184,7 +188,47 @@ class DebitCardPage extends Component {
...
+
+            // console.log('About to update state', expirationDate.substring(0, expirationDate.length - 2));
+            this.clearErrorAndSetValue('expirationDate', expirationDate.substring(0, expirationDate.length - 2));
+        } else {
+            this.allowExpirationDateChange = true;
+        }
+    }
+
     render() {
+        // console.log('Current State', this.state.expirationDate);
```
Comment (parasharrajat) [inline src/pages/settings/Payments/AddDebitCardPage.js:231]:
```suggestion
```

File: src/pages/settings/Payments/AddDebitCardPage.js [hunk 2]
```diff
@@ -184,7 +188,47 @@ class DebitCardPage extends Component {
...
+     */
+    removeSlashFromExpiryDate(e) {
+        const expirationDate = this.state.expirationDate;
+
+        // Auto remove the slash and digit when backspace is pressed and it ends with slash
+        if (e.nativeEvent.key === 'Backspace' && (expirationDate.length === 4 || (expirationDate.length === 3 && lodashEndsWith(expirationDate, '/')))) {
+            this.allowExpirationDateChange = false;
+            e.preventDefault();
+
+            // console.log('About to update state', expirationDate.substring(0, expirationDate.length - 2));
```
Comment (parasharrajat) [inline src/pages/settings/Payments/AddDebitCardPage.js:223]:
```suggestion
```

File: src/pages/settings/Payments/AddDebitCardPage.js [hunk 3]
```diff
@@ -184,7 +188,47 @@ class DebitCardPage extends Component {
...
+            expiryDate = `${inputExpiryDate.slice(0, 2)}/${inputExpiryDate.slice(2)}`;
+        }
+        this.clearErrorAndSetValue('expirationDate', expiryDate);
+    }
+
+    /**
+     * Remove slash from expiry date on backspace
+     * @param {Object} e
+     */
+    removeSlashFromExpiryDate(e) {
```
Comment (parasharrajat) [inline src/pages/settings/Payments/AddDebitCardPage.js:215]:
```suggestion
    removeSlashFromExpiryDate(event) {
```

File: src/pages/settings/Payments/AddDebitCardPage.js [hunk 4]
```diff
@@ -184,7 +188,47 @@ class DebitCardPage extends Component {
...
+        } else if (inputExpiryDate.length > 2 && _.indexOf(inputExpiryDate, '/') === -1) {
+            // Expiry Date with MM and YY without slash, hence adding slash(/)
+            expiryDate = `${inputExpiryDate.slice(0, 2)}/${inputExpiryDate.slice(2)}`;
+        }
+        this.clearErrorAndSetValue('expirationDate', expiryDate);
+    }
+
+    /**
+     * Remove slash from expiry date on backspace
+     * @param {Object} e
```
Comment (parasharrajat) [inline src/pages/settings/Payments/AddDebitCardPage.js:213]:
```suggestion
     * @param {Object} event
```

File: src/pages/settings/Payments/AddDebitCardPage.js [hunk 5]
```diff
@@ -184,7 +188,47 @@ class DebitCardPage extends Component {
         }));
     }
 
+    /**
+     * Add slash to the expiry date
```
Comment (parasharrajat) [inline src/pages/settings/Payments/AddDebitCardPage.js:192]:
```suggestion
```

File: src/pages/settings/Payments/AddDebitCardPage.js [hunk 6]
```diff
@@ -184,7 +188,47 @@ class DebitCardPage extends Component {
...
+            expiryDate = `${inputExpiryDate}/`;
+        } else if (inputExpiryDate.length > 2 && _.indexOf(inputExpiryDate, '/') === -1) {
+            // Expiry Date with MM and YY without slash, hence adding slash(/)
+            expiryDate = `${inputExpiryDate.slice(0, 2)}/${inputExpiryDate.slice(2)}`;
+        }
+        this.clearErrorAndSetValue('expirationDate', expiryDate);
+    }
+
+    /**
+     * Remove slash from expiry date on backspace
```
Comment (parasharrajat) [inline src/pages/settings/Payments/AddDebitCardPage.js:212]:
```suggestion
```

**Review by parasharrajat — CHANGES_REQUESTED** [review]

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Could you please test it carefully?

Android: 
Steps: 
1. Add a full date value. 
2. Start clearing the value until you reach the first char for a year. 
3. Remove one more char. Now value would be MM
4. try adding a new value for YYYY.

https://user-images.githubusercontent.com/24370807/159025466-6b93be19-c1fd-40ce-a921-1b2095b0f405.mp4

**Comment by mananjadhav** [comment]
@parasharrajat  I did test this case on iOS and other platforms but guess missed it on Android. I feel we should take your suggestion of using only `onChangeText`. I checked that the code doesn't work well on Android if I use the hardware keypad.

```
addOrRemoveSlashToExpiryDate(inputExpiryDate) {
        let expiryDate = inputExpiryDate;
        // Backspace was hit/character removed
        if (inputExpiryDate.length < this.state.expirationDate.length
          && (((inputExpiryDate.length === 3 && lodashEndsWith(inputExpiryDate, '/'))
          || (inputExpiryDate.length === 2 && lodashEndsWith(this.state.expirationDate, '/'))))) {
            expiryDate = inputExpiryDate.substring(0, inputExpiryDate.length - 1);
        } else if (inputExpiryDate.length === 2 && _.indexOf(inputExpiryDate, '/') === -1) { 
            // Expiry Date(MM)is added so append a slash(/)
            expiryDate = `${inputExpiryDate}/`;
        } else if (inputExpiryDate.length > 2 && _.indexOf(inputExpiryDate, '/') === -1) {
            // Expiry Date with MM and YY without slash, hence adding slash(/)
            expiryDate = `${inputExpiryDate.slice(0, 2)}/${inputExpiryDate.slice(2)}`;
        }
        this.clearErrorAndSetValue('expirationDate', expiryDate);
    }
```

**Comment by parasharrajat** [comment]
Could you please explain to me what is you trying on achieve in the code snippet above?

**Comment by mananjadhav** [comment]
There are two parts to the above snippet. Because `onKeyPress` isn't working well in Android, I am handling the backspace in the onChangeText only.

1. if the current text length is less than the previous value (in state), we treat that as backspace
      a. `(inputExpiryDate.length === 3 && lodashEndsWith(inputExpiryDate, '/')` Case when string was `22/2` and we pressed  backspace
      b. `inputExpiryDate.length === 2 && lodashEndsWith(this.state.expirationDate, '/')` Case when the string was `22/` and we press backspace
2. else the previous code of adding slash based on the text length

I can move the if conditions to a separate function to make it readable.

**Comment by parasharrajat** [comment]
I am trying to build the app on IOS. There are some issues with my MAC.

File: src/pages/settings/Payments/AddDebitCardPage.js [hunk 7]
```diff
@@ -91,6 +92,8 @@ class DebitCardPage extends Component {
         this.submit = this.submit.bind(this);
         this.clearErrorAndSetValue = this.clearErrorAndSetValue.bind(this);
         this.getErrorText = this.getErrorText.bind(this);
+        this.allowExpirationDateChange = true;
```
Comment (parasharrajat) [inline src/pages/settings/Payments/AddDebitCardPage.js:95]: Are we using this anywhere?

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Sorry for the delay in review. Facing a lot of issues on Mac.

Tested and it looks fine to me. I felt a little delay on Android. @Julesssss Could you please test this one?

**Comment by mananjadhav** [comment]
>  I felt a little delay on Android.

I tested on the emulator (don't have an Android device at the moment) and yes I did find the delay. I found the delay even with the previous backspace implementation. Did you try on an emulator or a device?

**Comment by parasharrajat** [comment]
I tried an emulator but I think it affects the device in the same way.

**Comment by mananjadhav** [comment]
I couldn't find a specific issue with the code and the same delay isn't experienced in iOS.

> Could it be that we're unnecessarily re-rendering the entire Add debit card form

It does rerender the component, should we try moving this into a separate TextInput? 

@Julesssss  I'll be able to access an Android device by tomorrow, can test release build after that.

**Comment by parasharrajat** [comment]
Great. Thanks, @mananjadhav. Let's wait a bit for the slack thread.

**Comment by parasharrajat** [comment]
@Julesssss What do you think would be the best path forward here?

**Comment by Julesssss** [comment]
Sorry, I think my Slack thread comment was about a different issue 🤦 

Maybe we continue with the release build test. If there's no noticeable lag we can probably just move forward.

> I'll be able to access an Android device by tomorrow, can test release build after that.

**Comment by mananjadhav** [comment]
> Maybe we continue with the release build test. If there's no noticeable lag we can probably just move forward.

@parasharrajat @Julesssss I was able to get access to an Android device and test a release build. I don't see a huge lag and it looks good enough to move forward. Attached is the video.

https://user-images.githubusercontent.com/3069065/163016741-ad84b11f-7377-4813-913d-97402dc5c92c.mp4

File: src/pages/settings/Payments/AddDebitCardPage.js (hunk 7 above)
Comment (mananjadhav) [reply src/pages/settings/Payments/AddDebitCardPage.js]: Removed

**Comment by Julesssss** [comment]
> @parasharrajat @Julesssss I was able to get access to an Android device and test a release build. I don't see a huge lag and it looks good enough to move forward. Attached is the video.

Awesome, thanks for that! Looks fine to me.

File: src/pages/settings/Payments/AddDebitCardPage.js [hunk 8]
```diff
@@ -184,6 +186,36 @@ class DebitCardPage extends Component {
...
 
+    /**
+     * @param {String} inputExpiryDate
+     */
+    addOrRemoveSlashToExpiryDate(inputExpiryDate) {
+        this.setState((prevState) => {
+            let expiryDate = inputExpiryDate;
+
+            // Backspace was hit/character removed
+            // Remove extra digit when there's a / involved
```
Comment (Julesssss) [inline src/pages/settings/Payments/AddDebitCardPage.js:197]: Could you just reword this sightly? It took me a couple of times to understand it properly

Comment (mananjadhav) [reply src/pages/settings/Payments/AddDebitCardPage.js]: I've updated. Let me know if it is fine.

File: src/pages/settings/Payments/AddDebitCardPage.js [hunk 9]
```diff
@@ -184,6 +186,36 @@ class DebitCardPage extends Component {
...
+            if (inputExpiryDate.length < prevState.expirationDate.length
+                && (((inputExpiryDate.length === 3 && lodashEndsWith(inputExpiryDate, '/'))
+                  || (inputExpiryDate.length === 2 && lodashEndsWith(prevState.expirationDate, '/'))))) {
+                expiryDate = inputExpiryDate.substring(0, inputExpiryDate.length - 1);
+            } else if (inputExpiryDate.length === 2 && _.indexOf(inputExpiryDate, '/') === -1) {
+                // Expiry Date(MM)is added so append a slash(/)
```
[the remark this answers is not in the document]
Comment (mananjadhav) [reply src/pages/settings/Payments/AddDebitCardPage.js]: Updated

**Review by Julesssss — APPROVED** [review]
Thanks. Looking good

**Review by parasharrajat — APPROVED** [review]
LGTM. 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Pull request merged**